src/libs/ajax/DatasetBuilder.ts

```diff
@@ -1,11 +1,93 @@
+// Types that can be used to create a criteria.
+
+export interface DomainType {
+  id: number;
+  category: string;
+  values: string[];
+}
+
+type DataType = 'range' | 'list';
+
+export interface ProgramDataType {
+  id: number;
+  name: string;
+  dataType: DataType;
+}
+
+export interface ProgramDataRangeType extends ProgramDataType {
+  min: number;
+  max: number;
+}
+
+export interface ProgramDataListTypeValue {
+  id: number;
+  name: string;
+}
+
+export interface ProgramDataListType extends ProgramDataType {
+  values: ProgramDataListTypeValue[];
+}
+
+// discriminating union
+// define the three types
+export type CriteriaType = DomainType | ProgramDataRangeType | ProgramDataListType;
```

Comment:
taking a quick look at how these are used, it looks like CriteriaType union is not used anywhere... and DomainType appears separated in use from the other two subtypes.  Maybe the right union to export is just type ProgramDataType = ProgramDataRangeType | ProgramDataListType, which would see nice usage on line 37 below and maybe elsewhere.

Reply:
`CriteriaType` is used in `CohortEditor.ts`; I've since moved it so it's next to its usage, since it isn't used here.

Regarding the type union, the type `ProgramDataRangeType | ProgramDataListType` is only used on line 37 so I don't think it's helpful to make a named type for it. The name `ProgramDataType` already exists as the base interface that both `ProgramDataListType` and `ProgramDataRangeType` share, so it would need to have a different name.

I was hoping to use a "base" interface type to represent this concept, like you would in Java, but since these are all interfaces it's easy to make an "instance" of the "abstract" base type by accident, which I want to disallow.

Reply:
ah... I missed that.  Ok that works.